presto-hive/src/test/java/com/facebook/presto/hive/TestHiveRecoverableGroupedExecution.java

```diff
@@ -282,79 +307,110 @@ public void testScanFilterProjectionOnlyQueryOnUnbucketedTable()
+            // test no failure case
+            Stopwatch noRecoveryStopwatch = Stopwatch.createStarted();
+            assertEquals(queryRunner.execute(recoverableSession, queryWithoutFailure).getUpdateCount(), OptionalLong.of(expectedUpdateCount));
+            log.info("Query with no recovery took %sms", noRecoveryStopwatch.elapsed(MILLISECONDS));
 
-                // test failure case
-                ListenableFuture<MaterializedResult> result = executor.submit(() -> queryRunner.execute(recoverableSession, queryWithFailure));
+            // cancel all queries and tasks to make sure we are dealing only with a single running query
+            cancelAllQueries(queryRunner);
+            cancelAllTasks(queryRunner);
 
-                // wait for split scheduling starts
-                while (!isSplitSchedulingStarted(queryRunner)) {
-                    sleep(10);
-                }
+            // test failure case
+            Stopwatch recoveryStopwatch = Stopwatch.createStarted();
+            ListenableFuture<MaterializedResult> result = executor.submit(() -> queryRunner.execute(recoverableSession, queryWithFailure));
 
-                // wait for additional 300ms before close a worker
-                sleep(300);
+            List<TestingPrestoServer> workers = queryRunner.getServers().stream()
+                    .filter(server -> !server.isCoordinator())
+                    .collect(toList());
+            shuffle(workers);
 
-                assertTrue(getRunningQueryId(queryRunner.getQueries()).isPresent());
-                TestingPrestoServer worker = queryRunner.getServers().stream()
-                        .filter(server -> !server.isCoordinator())
-                        .findFirst()
-                        .get();
-                worker.close();
+            TestingPrestoServer worker1 = workers.get(0);
+            // kill worker1 right away, to make sure recoverable execution works in cases when the task hasn't been yet submitted
+            worker1.stopResponding();
 
-                assertEquals(result.get(30, SECONDS).getUpdateCount(), OptionalLong.of(expectedUpdateCount));
-            }
-            finally {
-                for (@Language("SQL") String postQuery : postQueries) {
-                    queryRunner.execute(recoverableSession, postQuery);
+            // kill worker2 only after the task has been scheduled
+            TestingPrestoServer worker2 = workers.get(1);
+            Stopwatch stopwatch = Stopwatch.createStarted();
+            while (true) {
+                // wait for a while
+                sleep(500);
+
+                // if the task is already running - move ahead
+                if (hasTaskRunning(worker2)) {
+                    break;
+                }
+
+                // it is possible that the task has finished it's execution already
+                // don't fail the test
+                if (stopwatch.elapsed(SECONDS) > 5) {
```

Comment:
Curious: When would this happen?

Reply:
For example if for some reason (in case of a very miss-fortunate scheduling) the thread pooling for the task status is not scheduled before the task finishes.